### ISSUE TITLE
Mock scripts using match / replace

### ISSUE THREAD

**Issue by kleijnweb** [firecow/gitlab-ci-local#601]
It would be extremely handy if gcl was able to mock lines of `script`, by replacing it with a local script when it matches a certain regular expression.

Right now the only way to mock a feature that is actually from an image, is override a variable that hold the image tag, push a different version with different behavior, and then use that. That is doable but not ideal.

**Comment by firecow** [comment]
Please elaborate.

I don't quite understand the feature request.

**Comment by bcouetil** [comment]
There is sometimes situations when we don't want do to the same commands locally.

For these cases, I use $GITLAB_CI variable, which is "true" only in CI.

````
- if [ "$GITLAB_CI" == "true" ]; then ...
```

Would it solve your problem ?

**Comment by kleijnweb** [comment]
Lets say you have this in a job:

```
scripts:
  - ./do-something.sh arg1 arg2
```
It would be handy if we could have something completely different executed, without having to modify it and check for the $GITLAB_CI variable.

Additionally and perhaps more importantly, if you have something like this:

```
image: someimage
scripts:
  - ./do-something-with-script-in-image.sh
```

Lets say that script does something you don't want to happen locally, perhaps connecting remote infrastructure. The image might be managed by a completely different department, making it hard to modify. You can create a new image that does what you want it to for testing, but that is a PITA. It would be a lot easier and cleaner if you had a script on the host machine that is mounted and then executed.

In general, it would be better to add as little customization needed purely for GCL to the GitLab templates.

Config in GCL for something like that could look like this:

```
script_mocks:
  shell:
    - { match: '^./do-something.sh arg1', run: '/some/path/do-something-else.sh foo bar' }
  docker:
    someimage:
      - { match: '^./do-something.sh arg1', run: '/some/path/on/host/which/will/be/mounted/do-something-else.sh foo bar' }
```

**Comment by firecow** [comment]
In the "docker image job" I would use gitlab-ci-local --volume to overwrite the files you need to overwrite.

**Comment by firecow** [comment]
In the shell job you are in complete control of the script being called, and should be able to GITLAB_CI your way out of any dev specifics.

**Comment by firecow** [comment]
A "custom" object called scripts_mock would not pass the Gitlab CI/CD linter on your Gitlab instance.

**Comment by kleijnweb** [comment]
> A "custom" object called scripts_mock would not pass the Gitlab CI/CD linter on your Gitlab instance, as the json schema would be violated

I was not proposing that.

> If you need to override files outside of CI_PROJECT_DIR like /use/local/bin/stuff.sh in docker images the --volume option can be used. (Cli options can live in the .gitlab-ci-local-env file

Fair point. I was hoping for something a bit more robust and clear but I guess that will work for now.

**Comment by firecow** [comment]
If you want something "more robust".
I'll be glad to help.

If you paste a full example of a realistic use case you are puzzling with, it would be easier for us to understand what you are trying to achieve, and perhaps come up with some robust feature.

**Comment by kleijnweb** [comment]
More robust to me means:

- Not introducing mainline flow branching for testing
- Being explicit about the purpose of a piece of configuration

Relying on $GITLAB_CI for these type of things is sort of like having something like this in your production code:

```
if (TESTING) {
  doMockOfThing();
} else {
  doThing();
}
```

That to me is *not great*(TM). You would solve this by injecting mocks, either direct or via DI container, or whatever, not important. The important thing is that your testing code / logic is separate from your production code.

So for docker executor jobs, yes, you *could* mount files to override binaries and scripts in the target image, using the `VOLUME` env. I had not considered that, and that *may* solve my immediate problem. And unlike using $GITLAB_CI, that is neatly separate from the "production code", if you can call it that.

But, the `VOLUME` option was not created with that purpose in mind (correct me if I'm wrong), so it's not immediately obvious to anyone looking at it after me, why it is there. Resulting in some head scratching. And I foresee it getting messy very quickly if you want to mount over a lot of binaries and scripts, in a lot of jobs. Also what if you have multiple images where the same file is actually a different binary or script. For example perhaps different versions? The volume argument is applied to every service, so there's no way to distinct. Different arguments? I guess I'm parsing arguments in my mock script now.

For my specific use case, right now, I only need to override a single script, in a single image. So for me right now, using `VOLUME` is acceptable.

So to summarize, my ideal solution would:

- Not introduce any changes to the GitLab config
- Have configuration that has obvious semantics
- Allow easy override of whole lines of `script`, even if they use multiple binaries and shell scripts
- Allow fine grained matching of script and context, at least matching different images and shell, and ideally using regex matching for the script line (for example mock `./foo A` but not `./foo B`, etc)

**Comment by firecow** [comment]
I too hate if (prod) stuff.
GITLAB_CI == false/true resembles that, I totally agree.

I also agree that --volume wasn't created for this purpose.

What we need is a detailed feature request, a very precise solution
, and it will most definitely be taken seriously 👍

**Comment by kleijnweb** [comment]
I just stumbled on this project yesterday and am no expert on the inner workings of GLCI so I don't think I'm the right person to provide that precision, while taking into account how things currently work.

Right now I'm actually more inclined to fork gitlab-runner to make it easier to use for local testing. No idea how feasible that is, have you looked into that?

**Comment by firecow** [comment]
gitlab-runner is runnable with very limited CLI, and is not capable of interpreting a vast amount of .gitlab-ci.yml instructions.
There is a big issue on gitlab's forum about make it better, or coming up with some other solution to make .gitlab-ci.yml runnable locally.
